src/parser.rs

```diff
@@ -166,51 +155,244 @@ impl<'a> DocParser<'a> {
-  pub fn parse_module(
-    &self,
-    specifier: &ModuleSpecifier,
-  ) -> Result<ModuleDoc, DocError> {
-    let module_info = self.get_module_info(specifier)?;
-    let definitions = self.get_doc_nodes_for_module_info(module_info)?;
-    self.collect_diagnostics_for_nodes(&definitions);
-    let reexports = self.get_reexports_for_module(module_info);
-    let module_doc = ModuleDoc {
-      definitions,
-      reexports,
-    };
-    Ok(module_doc)
-  }
-
   fn get_module_info(
     &self,
     specifier: &ModuleSpecifier,
   ) -> Result<ModuleInfoRef, DocError> {
     get_module_info(&self.root_symbol, specifier)
   }
 
-  /// Fetches `file_name` and returns a list of exported items (no reexports).
-  #[cfg(feature = "rust")]
   pub fn parse(
     &self,
+  ) -> Result<IndexMap<ModuleSpecifier, Vec<DocNode>>, DocError> {
+    let mut doc_nodes_by_url = self
+      .specifiers
+      .iter()
+      .map(|specifier| {
+        Ok((
+          specifier.clone(),
+          self.parse_with_reexports_inner(specifier, HashSet::new())?,
+        ))
+      })
+      .collect::<Result<IndexMap<_, _>, DocError>>()?;
+
+    let mut all_locations = doc_nodes_by_url
+      .values()
+      .flatten()
+      .flat_map(|node| {
+        fn walk_rc_nodes(nodes: &[Rc<DocNode>]) -> Vec<Location> {
+          nodes
+            .iter()
+            .flat_map(|node| {
+              if let Some(namespace) = node.namespace_def() {
+                walk_rc_nodes(&namespace.elements)
+              } else {
+                vec![node.location.clone()]
+              }
+            })
+            .collect()
+        }
+
+        if let Some(namespace) = node.namespace_def() {
+          walk_rc_nodes(&namespace.elements)
+        } else {
+          vec![node.location.clone()]
+        }
+      })
+      .collect::<HashSet<_>>();
+
+    for (specifier, nodes) in doc_nodes_by_url.iter_mut() {
+      self.resolve_references_for_nodes(
+        specifier,
+        nodes,
+        &[],
+        &mut all_locations,
+      )?;
+    }
+
+    for (_, doc_nodes) in &doc_nodes_by_url {
+      self.collect_diagnostics_for_nodes(doc_nodes);
+    }
+
+    Ok(doc_nodes_by_url)
+  }
+
+  fn resolve_references_for_nodes(
+    &self,
     specifier: &ModuleSpecifier,
-  ) -> Result<Vec<DocNode>, DocError> {
-    let module_info = self.get_module_info(specifier)?;
-    let doc_nodes = self.get_doc_nodes_for_module_info(module_info)?;
-    self.collect_diagnostics_for_nodes(&doc_nodes);
-    Ok(doc_nodes)
+    nodes: &mut Vec<DocNode>,
+    name_path: &[String],
+    all_locations: &mut HashSet<Location>,
+  ) -> Result<(), DocError> {
+    let mut i = 0;
+    while i < nodes.len() {
+      let name = nodes[i].name.to_string();
+
+      match &mut nodes[i].def {
+        DocNodeDef::Namespace { namespace_def } => {
+          let mut namespace_elements = namespace_def
+            .elements
+            .iter()
+            .map(|node| Rc::unwrap_or_clone(node.clone()))
```

Comment:
`unwrap_or_clone` will always clone? Maybe just do:

```suggestion
            .map(|node| node.as_ref().clone())
```